### DISCUSSION TITLE
SQLArchemy with asyncio

### DISCUSSION THREAD

**Discussion by bert2002** [fastapi-users/fastapi-users#975] [Q&A] [answered]
Hi,
I am currently migrating from `8.1.3` to `9.3.2` and the database adapter changed a lot. I am trying to use mysql with it, but when I try to start my application I am getting:
```
sqlalchemy.exc.InvalidRequestError: The asyncio extension requires an async driver to be used. The loaded 'mysqldb' is not async.
```
and I checked and apparently it is possible to use it with `mysql+aiomysql` for the connection string, but it seems I cannot pass the supported scheme for the backend and run in e.g. 
```
    backend_str = self.SUPPORTED_BACKENDS[self.url.scheme]
KeyError: 'mysql+aiomysql'
```

Has anyone got this to work with mysql?

bert

**Comment by frankie567** [comment, marked as answer]
Did you install the `aiomysql` driver (`pip install aiomysql`)? I have another project which is able to run with MySQL without any particular issue, as long as the connection string is defined with the right scheme `mysql+aiomysql://foo:bar@localhost:3306/db`

**Reply by bert2002** [reply]
Yes installed the driver
```
# pip3 list |grep aiomysql
aiomysql                    0.0.21     
```
and the scheme should be correct: `mysql+aiomysql://user:password@localhost:3306/dbname`.
Anything that needs to be loaded extra?

**Reply by frankie567** [reply]
Hmmm... Do you have SQLAlchemy 1.4+ installed?

**Reply by bert2002** [reply]
Seems like it
```
# pip3 list |grep -i sqlal
fastapi-users-db-sqlalchemy 3.0.1               
SQLAlchemy                  1.4.23    
```

**Reply by frankie567** [reply]
And your `fastapi-users-db-sqlalchemy >= 2.0.0`? 😅

**Reply by bert2002** [reply]
ahh there is still some `database` loaded:
```
    database = databases.Database(os.environ.get('DATABASE_URL'))
  File "/home/user/.local/lib/python3.8/site-packages/databases/core.py", line 65, in __init__
    backend_str = self.SUPPORTED_BACKENDS[self.url.scheme]
KeyError: 'mysql+aiomysql'
```
I guess that is confusing it

**Reply by frankie567** [reply]
Yeah, you should completely get rid of `encode/databases` with this newer version

**Comment by caniko** [comment]
Does this mean we can't use `trio` when using async? I am not sure how well these work together.

**Reply by frankie567** [reply]
The SQLAlchemy database adapter entirely relies on the async support developed by SQLAlchemy (version 1.4+): https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html

I think they use [`greenlet`](https://pypi.org/project/greenlet/) to manage async threads. I'm not sure it's possible to select another implementation like `trio`.
